pytorch_lightning/metrics/functional/psnr.py

```diff
@@ -58,8 +73,13 @@ def psnr(
 
     """
     if data_range is None:
+        if dim is not None:
+            # Maybe we could use `torch.amax(target, dim=dim) - torch.amin(target, dim=dim)` in PyTorch 1.7 to calculate
+            # `data_range` in the future.
+            raise ValueError("`data_range` must be given when `dim` is not `None`.")
+
         data_range = target.max() - target.min()
     else:
         data_range = torch.tensor(float(data_range))
```

Comment:
could we move all this logic into the `_update` function?

Reply:
In the original implementation (before this PR) and the updated commits (after the comment), the `data_range` is passed to `functional._psnr_compute`, maybe we could move the checks on `data_range` and `dim` from `functional.psnr` and `PSNR.__init__` to `_compute`. But the users of `PSNR` will get the exception until they call `PSNR.compute` (or `on_*_epoch_end` is called).

Should we extract the logic to a single place (`functional._psnr_compute`)?